python/ray/data/_internal/datasource/parquet_datasource.py

```diff
@@ -441,10 +452,108 @@ def get_current_projection(self) -> Optional[List[str]]:
+            partition columns), or [] if projection pushdown excluded partition
+            columns (meaning include no partition columns).
+        """
+        if self._projection_map is None:
+            return None
+
+        partition_cols_set = self._partition_columns_set
+        if not partition_cols_set:
+            return None
+
+        # Extract partition columns that are in the projection map
+        partition_cols = [
+            col for col in self._projection_map.keys() if col in partition_cols_set
+        ]
+
+        # If partition columns are found in projection map, return them
+        if partition_cols:
+            return partition_cols
+
+        # No partition columns in projection map:
+        # - If _partition_columns is None: user only specified data columns during
+        #   initialization, so include all partition columns (return None)
+        # - If _partition_columns is not None: partition columns were requested during
+        #   initialization but are missing from _projection_map, which means projection
+        #   pushdown excluded them, so exclude all partition columns (return [])
+        return None if self._partition_columns is None else []
+
+    def _get_data_columns(self) -> Optional[List[str]]:
+        """Extract data columns from projection map, excluding partition columns.
+
+        Partition columns aren't in the physical file schema, so they must be
+        filtered out before passing to PyArrow's to_batches().
+
+        Returns:
+            List of data column names to read from files, or None if no projection.
+            Can return empty list if only partition columns are projected.
+        """
+        if self._projection_map is None:
             return None
 
-        return (data_columns or []) + (self._partition_columns or [])
+        # Get partition columns and filter them out from the projection
+        partition_cols = self._partition_columns_set
+        data_cols = [
+            col for col in self._projection_map.keys() if col not in partition_cols
+        ]
+
+        return data_cols
+
+    def apply_predicate(
+        self,
+        predicate_expr: Expr,
+    ) -> "ParquetDatasource":
+        """Apply a predicate to this datasource, separating partition column predicates.
+
+        Partition column predicates aren't pushed down to PyArrow since partition
+        columns aren't in the physical file schema. If a predicate references partition
+        columns, we return self unchanged so the Filter operator remains in the plan
+        and can be applied after partition columns are added.
+        """
+        # Check if predicate references any partition columns
+        referenced_cols = set(get_column_references(predicate_expr))
+        partition_cols = self._partition_columns_set
+
+        if referenced_cols & partition_cols:
+            # Don't push down predicates on partition columns
+            return self
+
```

Comment:
Dumb question -- can't we still pushdown the data columns, and just filter the partition values in the reader implementation? Like, is it necessary to disable predicate pushdown altogether?